### ISSUE TITLE
PDFBoxProcessor fails to read metadata from some PDFs

### ISSUE THREAD

**Issue by cmhdave** [cantaloupe-project/cantaloupe#366]
Before Cantaloupe we provided images of PDF pages via Ghostscript and didn't have any problems.  Once we started testing these PDFs with Cantaloupe and Apache PDFBox, many of the PDFs that were working fine with Ghostscript started failing in Cantaloupe.  

We could resolve the issue by manually optimizing the PDFs with Ghostscript but we wouldn't necessarily have that option in a large scale (nor find all the PDFs that need optimized).  

Now I realize this is NOT a Cantaloupe issue, rather it's an Apache PDFBox issue and the direction to take to get that resolved would be to open a ticket with that project with an offending PDF to see if they can resolve it so that Cantaloupe can just upgrade its dependency.

However as I was reading through Apache PDFBox's Issue log it seems there are many different rendering issues that they are constantly dealing with as they come up.  On the other hand we know that Ghostscript handles these PDFs already without a hitch.  So my question here is for providing a new Processor that uses Ghost4J to process PDFs and users of Cantaloupe could decide to use one or the other (much the same as you have conveniently done for other image formats)

This is certainly not a high priority bug... it's a requested improvement.  I also would be curious if this even fits in your long term vision for the project. (This kind of interoperability of different processors for different image types)

**Comment by kaij** [comment]
Rendering PDFs in Cantaloupe is maybe not the best thing (Memory, CPU, ...) What I am doing is rendering PDFs in a HTTP microservice outside of cantaloupe and requesting them through cantaloupe using HTTPSource. This frees resources from cantaloupe and at the same time leverages source and derivative caching in cantaloupe.

**Comment by adolski** [comment]
I looked into Ghost4J a while back and I remember reading that it's not thread-safe, as in multiple threads in the same process can't read different files at the same time. I don't know if that is correct, or if it's a limitation of Ghost4J or Ghostscript, but it definitely won't work in a Servlet situation.

I know that Ghostscript also has a command-line tool, which would work around the thread safety issue, but these have certain limitations and it's better to use in-process API if possible (my opinion).

I don't disagree about the PDFBox issues. PDFBoxProcessor exists basically because PDFBox was free and easy to incorporate. I'm not aware of all of the other free PDF libraries that are available. I know of [Poppler](https://en.wikipedia.org/wiki/Poppler_(software)) but the challenge of this and similar libraries is the bridge from Java into native code, which requires JNI or JNA or JNR or waiting for Project Panama or whatever (and I'm not familiar with any of those and don't have time to learn them right now).

As far as long-term vision: I would like to move format support into Image I/O plugins--to have fewer processors and more IIO plugins. If you have the development resources, and you want to utilize some other codec like Poppler, you could develop a standard Java Image I/O plugin, and then that plugin should work in Cantaloupe with only very minor changes.

**Comment by cmhdave** [comment]
Ghost4J supports multi-threading as of version 0.4.0 but their explanation of how it's done seems kind of clunky.  

I love your idea of an Image I/O plugin.  I'll investigate that further.

**Comment by cmhdave** [comment]
I've tinkered with Ghost4J and have had mostly positive results with regard to performance and ability to open files that PdfBox cannot.  There are caveats though.  The way Ghost4J accomplishes multi-threaded processing is by running Ghostscript in a separate JVM.  I found that I had to make sure my JAVA_HOME was set properly on the servers I ran Cantaloupe on so that Ghost4J was able to run the JVM.  

I also chose to create a custom ImageRenderer for the Ghost4J library that enhances the way Ghostscript is called with correct number of processors (making Ghostscript multi-threaded as well as Ghost4J), increases the size of memory used by Ghostscript from 8MB to 30MB, and update the MaxPatternBitmap parameter.  These settings were recommended in this URL

https://www.ghostscript.com/doc/current/Use.htm#Improving_performance

I do this as a quick proof of concept to see if I could get a processor running that will open our PDF files where PdfBox fails.  I realize that this particular processor would introduce the need for a user to install another dependency in order to use and isn't standalone like PdfBox.  I also realize it is not using an Image I/O plugin which would be a longer term goal for this kind of change.
